oximeter/producer/src/lib.rs

```diff
@@ -274,21 +423,152 @@ pub async fn register(
+                } else {
+                    BackoffError::permanent(msg)
+                }
+            }
+            _ => BackoffError::permanent(e.to_string()),
+        })
+    };
+    backoff::retry_notify(
+        backoff::retry_policy_internal_service(),
+        do_register,
+        log_failure,
+    )
+    .await
+    .expect("Expected infinite retry loop registering as a producer with")
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Config;
+    use super::LogConfig;
+    use super::ProducerEndpoint;
+    use super::Server;
+    use dropshot::ConfigDropshot;
+    use httptest::matchers::request;
+    use httptest::responders::status_code;
+    use httptest::Expectation;
+    use httptest::Server as TestServer;
+    use omicron_common::api::internal::nexus::ProducerKind;
+    use omicron_common::api::internal::nexus::ProducerRegistrationResponse;
+    use slog::Drain;
+    use slog::Logger;
+    use std::time::Duration;
+    use tokio::time::Instant;
+    use uuid::Uuid;
+
+    fn test_logger() -> Logger {
+        let dec =
+            slog_term::PlainSyncDecorator::new(slog_term::TestStdoutWriter);
+        let drain = slog_term::FullFormat::new(dec).build().fuse();
+        let log =
+            Logger::root(drain, slog::o!("component" => "fake-cleanup-task"));
+        log
+    }
+
+    #[tokio::test]
+    async fn test_producer_registration_task() {
+        let log = test_logger();
+        let mut fake_nexus = TestServer::run();
+        slog::info!(
+            log,
+            "fake nexus test server listening";
+            "address" => ?fake_nexus.addr(),
+        );
+
+        const INTERVAL: Duration = Duration::from_secs(1);
+        let body = ProducerRegistrationResponse { lease_duration: INTERVAL };
+        fake_nexus.expect(
+            Expectation::matching(request::method_path(
+                "POST",
+                "/metrics/producers",
+            ))
+            .times(1..)
+            .respond_with(
+                status_code(201).body(serde_json::to_string(&body).unwrap()),
             ),
-        ))
+        );
+        let address = "[::1]:0".parse().unwrap();
+        let config = Config {
+            server_info: ProducerEndpoint {
+                id: Uuid::new_v4(),
+                kind: ProducerKind::Service,
+                address,
+                base_route: String::from("/collect"),
+                interval: Duration::from_secs(10),
+            },
+            registration_address: Some(fake_nexus.addr()),
+            dropshot: ConfigDropshot {
+                bind_address: address,
+                ..Default::default()
+            },
+            log: LogConfig::Logger(log),
+        };
+
+        // Ideally, we would check pretty carefully that there are exactly N
+        // registrations after N renewal periods. That's brittle, especially on
+        // a loaded system. Instead, we'll run for a number of intervals, and
+        // just check that we have more than one.
```

Comment:
Should the `.times(1..)` above be `.times(2..)` if we want to check "more than one"?

Reply:
Done in 2c0afef2ebf1321aeca50dffd86df08894436b39